### PR TITLE
Generic Mqtt Test Module: TRC HTTP Client

### PR DESCRIPTION
**Pull request by and-rewsmith**
The mqtt test module I made is the first rust test module we have, so we will need a client in order to report test status to the TestResultCoordinator. I have made a separate folder (rust-test-util) under our existing test folder. This test folder will contain all the rust e2e test utils we add in the future. This PR adds the the trc-client crate as the first util. 

This api is swagger-spec'd out so I tried using swagger-codegen to generate a rust client for this api. This didn't work because of a versioning mismatch between the futures dependency between the test module and the generated client code. I couldn't find a suitable alternative so I just made the client manually.

I am working on checkin for this test module and adding logic for the test module to use the trc-client in separate PRs.

### CODE REVIEW

File: test/modules/GenericMqttTester/src/message_channel.rs [hunk 1]
```diff
@@ -38,11 +39,13 @@ pub trait MessageHandler {
...
+    reporting_client: TestResultReportingClient,
+}
 
 impl ReportResultMessageHandler {
-    pub fn new() -> Self {
-        Self {}
+    pub fn new(reporting_client: TestResultReportingClient) -> Self {
+        Self { reporting_client }
     }
 }
```
Comment (and-rewsmith) [inline test/modules/GenericMqttTester/src/message_channel.rs:50]: I am implementing the changes to the test module itself in a separate PR.

**Comment by and-rewsmith** [comment]
We might want to move the client generation logic for e2e tests (added in a prior PR) from mqtt/mqtt-test-util into this new folder for rust e2e test utils.

File: test/rust-test-util/TestResultReportingClient/src/models/test_result.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+pub struct TestOperationResultDto {
+    #[serde(rename = "source")]
+    source: String,
+    #[serde(rename = "result")]
+    result: TestResult,
+    #[serde(rename = "type")]
+    _type: String,
+    #[serde(rename = "createdAt")]
+    created_at: String,
+}
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result.rs:18]: I suppose we could add tests here for serialize / deserialize

File: test/rust-test-util/TestResultReportingClient/src/models/test_result.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            .ok_or(de::Error::missing_field("sequence_number"))?;
+
+        let test_result = TestResult::new(
+            tracking_id.clone(),
+            batch_id.clone(),
+            sequence_number.parse::<u32>().map_err(|_| {
+                de::Error::invalid_type(
+                    Unexpected::Str("sequence number should not be a string"),
+                    &"sequence number as u32",
+                )
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result.rs:77]: Not too sure how the grammar should be for these Unexpected / Expected arguments

File: test/rust-test-util/TestResultReportingClient/src/models/test_result.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    #[serde(rename = "result")]
+    result: TestResult,
+    #[serde(rename = "type")]
+    _type: String,
+    #[serde(rename = "createdAt")]
+    created_at: String,
+}
+
+impl TestOperationResultDto {
+    pub fn new(source: String, result: TestResult, _type: String, created_at: String) -> Self {
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result.rs:21]: I think that created_at should come in as datetime.

File: test/rust-test-util/TestResultReportingClient/src/models/test_result.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TestOperationResultDto {
+    #[serde(rename = "source")]
+    source: String,
+    #[serde(rename = "result")]
+    result: TestResult,
+    #[serde(rename = "type")]
+    _type: String,
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result.rs:15]: We should probably make this an enum. I believe appropriate values are 0 to ~10.

File: test/rust-test-util/TestResultReportingClient/src/models/message_result.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                    })?,
+                );
+                Ok(test_result)
+            }
+        }
+        {
+            deserializer.deserialize_identifier(CustomVisitor)
+        }
+    }
+}
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/message_result.rs:82]: On second thought, I don't think we will need to deserialize but it is here now with tests so I am ok keeping it.

File: test/rust-test-util/TestResultReportingClient/src/models/test_result_dto.rs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    #[serde(rename = "type")]
+    _type: u8,
+    #[serde(rename = "createdAt")]
+    created_at: String,
+}
+
+impl TestOperationResultDto {
+    pub fn new(
+        source: String,
+        result: MessageTestResult,
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result_dto.rs:21]: We technically should support more than just MessageTestResult, but that is all I see us needing for the forseeable future.

File: test/rust-test-util/TestResultReportingClient/src/models/test_result_dto.rs [hunk 8]
```diff
@@ -0,0 +1,98 @@
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+
+use crate::MessageTestResult;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct TestOperationResultDto {
```
Comment (vadim-kovalyov) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result_dto.rs:7]: nit: pub(crate). No need to expose dtos

File: test/rust-test-util/TestResultReportingClient/src/models/message_result.rs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+            fn expecting(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
+                formatter.write_str("semicolon separated array")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<MessageTestResult, E>
+            where
+                E: de::Error,
+            {
+                let parts: Vec<String> = value.split(";").map(|part| part.to_string()).collect();
```
Comment (vadim-kovalyov) [inline test/rust-test-util/TestResultReportingClient/src/models/message_result.rs:56]: nit: will `.cloned().collect()` work?

File: test/rust-test-util/TestResultReportingClient/src/models/message_result.rs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                let tracking_id = parts
+                    .get(0)
+                    .ok_or(de::Error::missing_field("tracking_id"))?;
+                let batch_id = parts.get(1).ok_or(de::Error::missing_field("batch_id"))?;
+                let sequence_number = parts
+                    .get(2)
+                    .ok_or(de::Error::missing_field("sequence_number"))?;
+
+                let test_result = MessageTestResult::new(
+                    tracking_id.clone(),
```
Comment (vadim-kovalyov) [inline test/rust-test-util/TestResultReportingClient/src/models/message_result.rs:66]: unnecessary clone?

File: test/modules/GenericMqttTester/src/settings.rs [hunk 11]
```diff
@@ -36,6 +37,10 @@ impl Settings {
     pub fn test_scenario(&self) -> &TestScenario {
         &self.test_scenario
     }
+
+    pub fn test_result_coordinator_url(&self) -> &String {
+        &self.test_result_coordinator_url
+    }
 }
 
 impl<'de> serde::Deserialize<'de> for Settings {
```
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/modules/GenericMqttTester/src/settings.rs:46]: I don't think we will need to deserialize but it is here now with tests so I am ok keeping it.

File: test/rust-test-util/TestResultReportingClient/src/client.rs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            client: Client::new(),
+            uri,
+        }
+    }
+
+    pub async fn report_result(
+        &self,
+        source: String,
+        result: MessageTestResult,
+        _type: TestType,
```
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/rust-test-util/TestResultReportingClient/src/client.rs:33]: type was a reserved word, but I can change to test_type

File: test/rust-test-util/TestResultReportingClient/src/client.rs [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        test_result_dto::{TestOperationResultDto, TestType},
+    },
+    ReportResultError,
+};
+
+const CONTENT_TYPE: &str = "Content-Type";
+const APPLICATION_JSON: &str = "application/json";
+
+#[derive(Debug)]
+pub struct TestResultReportingClient {
```
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/rust-test-util/TestResultReportingClient/src/client.rs:16]: I agree, however there is already hyper::Client which will conflict. I'm going to change to TrcClient?

File: test/rust-test-util/TestResultReportingClient/src/models/message_result.rs (hunk 9 above)
Comment (and-rewsmith) [reply test/rust-test-util/TestResultReportingClient/src/models/message_result.rs:56]: It complains that Clone isn't implemented for str.

File: test/rust-test-util/TestResultReportingClient/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+[package]
```
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/rust-test-util/TestResultReportingClient/Cargo.toml:1]: This is a good point. I think it will be cleaner to make this change in a separate PR so we can keep the diff here more understandable.

File: test/modules/GenericMqttTester/src/settings.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/modules/GenericMqttTester/src/settings.rs:46]: I also think it is fine to remove, so lmk if you have a preference.

File: test/rust-test-util/TestResultReportingClient/src/models/test_result_dto.rs [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    Messages,
+    DirectMethod,
+    Twin,
+    Network,
+    Deployment,
+    EdgeHubRestartMessage,
+    EdgeHubRestartDirectMethod,
+    Error,
+    TestInfo,
+}
```
Comment (and-rewsmith) [inline test/rust-test-util/TestResultReportingClient/src/models/test_result_dto.rs:48]: Even though the client doesn't support all these test types, we need all of them here for the u8 conversion to be correct. I have logic to gate only supported test types in the client.

File: test/modules/GenericMqttTester/src/settings.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply test/modules/GenericMqttTester/src/settings.rs:46]: Let me rephrase - I'm not sure why couldn't you just `derive(Deserialize)` on `Settings` ? And if it is not used, I'd prefer to remove that.

**Review by vadim-kovalyov — DISMISSED** [review]
🚀

File: test/modules/GenericMqttTester/src/settings.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/modules/GenericMqttTester/src/settings.rs:46]: Oh I misread what this comment was under. I thought it was about the custom deserialize for MessageTestResult. This deserialize is used. I think you are right and I can just derive it. I was following the pattern in the bridge and didn't realize this.

[the remark this answers is not in the document]
Comment (and-rewsmith) [reply test/modules/GenericMqttTester/src/settings.rs:46]: I can remove the deserialize for MessageTestResult

**Review by vadim-kovalyov — APPROVED** [review]
